src/controller/proxy/controller.go

```diff
@@ -96,33 +101,55 @@ func (c *controller) UseLocalBlob(ctx context.Context, art lib.ArtifactInfo) boo
 	return exist
 }
 
-func (c *controller) UseLocalManifest(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface) (bool, error) {
+// ManifestList ...
+type ManifestList struct {
+	Content     []byte
+	Digest      string
+	ContentType string
+}
+
+func (c *controller) UseLocalManifest(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface) (bool, *ManifestList, error) {
 	a, err := c.local.GetManifest(ctx, art)
 	if err != nil {
-		return false, err
+		return false, nil, err
 	}
-	if a == nil {
-		return false, nil
+	// Pull by digest when artifact exist in local
+	if a != nil && len(art.Digest) > 0 {
+		return true, nil, nil
 	}
-	// Pull by digest
-	if len(art.Digest) > 0 {
-		return true, nil
-	}
-	// Pull by tag
+
 	remoteRepo := getRemoteRepo(art)
-	exist, dig, err := remote.ManifestExist(remoteRepo, art.Tag) // HEAD
+	exist, dig, err := remote.ManifestExist(remoteRepo, getReference(art)) // HEAD
 	if err != nil {
-		return false, err
+		return false, nil, err
 	}
 	if !exist {
 		go func() {
 			c.local.DeleteManifest(remoteRepo, art.Tag)
 		}()
-		return false, errors.NotFoundError(fmt.Errorf("repo %v, tag %v not found", art.Repository, art.Tag))
+		return false, nil, errors.NotFoundError(fmt.Errorf("repo %v, tag %v not found", art.Repository, art.Tag))
+	}
+
+	var content []byte
+	if c.cache != nil {
+		err = c.cache.Fetch(getManifestListKey(art.Repository, dig), &content)
+		if err == nil {
+			log.Debugf("Get the manifest list with key=cache:%v", getManifestListKey(art.Repository, dig))
+			return true, &ManifestList{content, dig, manifestlist.MediaTypeManifestList}, nil
+		}
+		if err == cache.ErrNotFound {
+			log.Debugf("Digest is not found in manifest list cache, key:cache:%v", getManifestListKey(art.Repository, dig))
```

Comment:
This is trivial but why key:cache:${key}?

Reply:
This is the actual key in the redis